### PR TITLE
Wrap new WebSocket connection in promise

### PR DESCRIPTION
**Pull request by niuyunsi**

### CODE REVIEW

File: src/data/ConferenceConnection.ts [hunk 1]
```diff
@@ -37,24 +40,31 @@ export interface MessageHandler {
...
         this.conn = conn;
-        this.conn.connect(url);
+        this.url = url;
         this.adapter = adapter;
+    }
 
-        this.conn.onmessage = (msg) => {
-            if (this.onConnMessageHandler) {
-                this.onConnMessageHandler(msg, () => {
+    public connect() {
```
Comment (linweigao) [inline src/data/ConferenceConnection.ts:54]: connect(): Promise<void>

File: src/data/ConferenceConnection.ts [hunk 2]
```diff
@@ -1,16 +1,19 @@
...
 export interface IConferenceConnection {
+    connect: () => Promise<any>;
     subscribe: (subscriber: ConferenceConnectionSubscriber) => void;
     publish: (message: IConfOutgoingMessage) => void;
     close: () => void;
 }
 
 export interface IConnection {
-    connect: (url: string) => void;
+    connect: (url: string) => Promise<any>;
```
Comment (linweigao) [inline src/data/ConferenceConnection.ts]: Promise<void>

File: src/components/Conference.tsx [hunk 3]
```diff
@@ -153,16 +153,20 @@ export class Conference extends React.Component<IConferenceProps, IConferenceSta
...
-        this.joinRoom(this.props.room);
-        this.getUserMedia().then(() => {
-            this.connection.subscribe(this.handleIncomingMessage)
-        });
+        this.connection = this.props.connect();
+
+        // NOTE(yunsi): Create websocket connection, if succeed then join room, if failed then fire error.
+        this.connection.connect()
+            .then(() => {
+                this.joinRoom(this.props.room);
```
Comment (linweigao) [inline src/components/Conference.tsx]: The logic here will be an issue.

joinRoom and getUserMedia should be Promise-Based API and put in the pipeline.

Comment (linweigao) [reply src/components/Conference.tsx]: Add a TODO here

File: src/components/Conference.tsx [hunk 4]
```diff
@@ -153,16 +153,20 @@ export class Conference extends React.Component<IConferenceProps, IConferenceSta
...
-        this.connection = connection
-        this.joinRoom(this.props.room);
-        this.getUserMedia().then(() => {
-            this.connection.subscribe(this.handleIncomingMessage)
-        });
+        this.connection = this.props.connect();
+
+        // NOTE(yunsi): Create websocket connection, if succeed then join room, if failed then fire error.
+        this.connection.connect()
+            .then(() => {
```
Comment (linweigao) [inline src/components/Conference.tsx:160]: then((), error handling should be here)

File: src/Spreed/SpreedConnection.ts [hunk 5]
```diff
@@ -17,15 +19,17 @@ export class SpreedConnection implements IConnection {
...
-        this.processRequests();
+        return new Promise((resolve, reject) => {
+            this.conn = new WebSocket(url)
+            this.conn.onmessage = this.onConnMessage.bind(this);
+            this.conn.onclose = this.onConnClose.bind(this);
+            this.conn.onopen = () => {
+                resolve()
+            };
+            this.conn.onerror = (err) => {
+                reject(err)
```
Comment (linweigao) [inline src/Spreed/SpreedConnection.ts:30]: Ideally here should give the error we defined.

File: src/Spreed/SpreedConnection.ts [hunk 6]
```diff
@@ -17,15 +19,17 @@ export class SpreedConnection implements IConnection {
...
-        this.conn = new WebSocket(url);
-        this.conn.onmessage = this.onConnMessage.bind(this);
-        this.conn.onclose = this.onConnClose.bind(this);
-        this.conn.onerror = this.onConnError.bind(this);
-        this.conn.onopen = this.onConnOpen.bind(this);
-    }
-
-    private onConnOpen() {
-        this.processRequests();
+        return new Promise((resolve, reject) => {
```
Comment (linweigao) [inline src/Spreed/SpreedConnection.ts]: We should make sure connect can only be called once. Check the status of conn.

File: src/components/Conference.tsx (hunk 4 above)
Comment (niuyunsi) [reply src/components/Conference.tsx:160]: Done

File: src/components/Conference.tsx (hunk 3 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Added TODO

File: src/data/ConferenceConnection.ts (hunk 2 above)
Comment (niuyunsi) [reply src/data/ConferenceConnection.ts]: Done

File: src/data/ConferenceConnection.ts (hunk 1 above)
Comment (niuyunsi) [reply src/data/ConferenceConnection.ts:54]: Done

**Review by linweigao — APPROVED** [review]

**Pull request merged**